File: algorithms/05_csharp/data_structures/circular_buffer.cs
using System;

namespace Algorithms.DataStructures
{
    public class CircularBuffer
    {
        private readonly int[] _data;
        private int _head;
        private int _size;

        public CircularBuffer(int capacity)
        {
            _data = new int[capacity];
        }

        public int Count => _size;

        public void Push(int value)
        {
            _data[(_head + _size) % _data.Length] = value;
            if (_size < _data.Length)
            {
                _size++;
            }
            else
            {
                _head = (_head + 1) % _data.Length;
            }
        }

        public int Pop()
        {
            int value = _data[_head];
            _head = (_head + 1) % _data.Length;
            _size--;
            return value;
        }

        public static void Main()
        {
            var buffer = new CircularBuffer(3);
            for (int i = 1; i <= 4; i++)
            {
                buffer.Push(i);
            }
            if (buffer.Pop() != 2 || buffer.Pop() != 3 || buffer.Pop() != 4)
            {
                throw new InvalidOperationException("bad overwrite semantics");
            }
            Console.WriteLine("circular buffer ok");
        }
    }
}

File: algorithms/04_java/data_structures/CircularBuffer.java
public final class CircularBuffer {
    private final int[] data;
    private int head = 0;
    private int size = 0;

    public CircularBuffer(int capacity) {
        data = new int[capacity];
    }

    public void push(int value) {
        data[(head + size) % data.length] = value;
        if (size < data.length) {
            size++;
        } else {
            head = (head + 1) % data.length;
        }
    }

    public int pop() {
        int value = data[head];
        head = (head + 1) % data.length;
        size--;
        return value;
    }

    public static void main(String[] args) {
        CircularBuffer buffer = new CircularBuffer(3);
        for (int i = 1; i <= 4; i++) {
            buffer.push(i);
        }
        if (buffer.pop() != 2 || buffer.pop() != 3 || buffer.pop() != 4) {
            throw new AssertionError("bad overwrite semantics");
        }
        System.out.println("circular buffer ok");
    }
}
